autogen/main.tf

```diff
@@ -36,290 +36,118 @@ resource "random_shuffle" "available_zones" {
+  location                    = var.regional ? var.region : var.zones[0]
+  node_locations              = var.regional ? sort(random_shuffle.available_zones.result) : slice(compact(var.zones), 1, length(compact(var.zones)))
+  latest_master_version       = data.google_container_engine_versions.master.latest_master_version
+  latest_node_version         = data.google_container_engine_versions.node.latest_master_version
+  kubernetes_master_version   = var.kubernetes_version != "latest" ? var.kubernetes_version : local.latest_master_version
+  kubernetes_node_version     = var.node_version != "" ? var.kubernetes_version : local.latest_node_version
   custom_kube_dns_config      = length(keys(var.stub_domains)) > 0
   upstream_nameservers_config = length(var.upstream_nameservers) > 0
   network_project_id          = var.network_project_id != "" ? var.network_project_id : var.project_id
   zone_count                  = length(var.zones)
-
-  cluster_type = var.regional ? "regional" : "zonal"
+  cluster_type                = var.regional ? "regional" : "zonal"
+  // auto upgrade by defasult only for regional cluster as long it has multiple masters versus zonal clusters have only have a single master so upgrades are more dangerous.
+  default_auto_upgrade = var.regional ? true : false
 
   cluster_network_policy = var.network_policy ? [{
     enabled  = true
     provider = var.network_policy_provider
-  }] : [{
-    enabled = false
+    }] : [{
+    enabled  = false
     provider = null
   }]
 
 {% if beta_cluster %}
-  cluster_cloudrun_config = var.cloudrun ? [{disabled = false}] : []
+  cluster_cloudrun_config = var.cloudrun ? [{ disabled = false }] : []
 
   cluster_node_metadata_config = var.node_metadata == "UNSPECIFIED" ? [] : [{
     node_metadata = var.node_metadata
   }]
 
 {% endif %}
-  cluster_type_output_name = {
-    regional = element(concat(google_container_cluster.primary.*.name, [""]), 0)
-    zonal = element(
-      concat(google_container_cluster.zonal_primary.*.name, [""]),
-      0,
-    )
-  }
-
-  cluster_type_output_location = {
-    regional = element(concat(google_container_cluster.primary.*.region, [""]), 0)
-    zonal = element(
-      concat(google_container_cluster.zonal_primary.*.zone, [""]),
-      0,
-    )
-  }
-
-  cluster_type_output_region = {
-    regional = element(concat(google_container_cluster.primary.*.region, [""]), 0)
-    zonal    = var.region
-  }
-
-  cluster_type_output_regional_zones = flatten(google_container_cluster.primary.*.node_locations)
-  cluster_type_output_zonal_zones    = local.zone_count > 1 ? slice(var.zones, 1, local.zone_count) : []
 
-  cluster_type_output_zones = {
-    regional = local.cluster_type_output_regional_zones
-    zonal = concat(
-      google_container_cluster.zonal_primary.*.zone,
-      local.cluster_type_output_zonal_zones,
-    )
-  }
+  cluster_output_name           = concat(google_container_cluster.primary.*.name, [""])[0]
```

Comment:
Why is the concat + indexing needed here. Since there's only a single resource now, can we reference it directly? This applies to all below.